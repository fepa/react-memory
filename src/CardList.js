import React from 'react';
import Card from './Card';

class CardList extends React.Component {
  constructor(props) {
    super(props);
    this.state = { order: [] };
    for(let i = 0; i < (this.props.count * 2); i++) {
      this.state.order.push(i);
    }
    this.shuffle(this.state.order);
  }

  matchesPreviousCard(card) {
    return this.props.activeCards[0].matches(card);
  }

  checkMatch(card) {
    // eslint-disable-next-line
    if(this.props.activeCards.length != 1) {
      this.props.setActive(card);
      return;
    }

    if(this.matchesPreviousCard(card)) {
      this.props.setMatch(card.props.pairing);
      this.props.clearActive();
    } else {
      this.props.setActive(card);
    }
  }

  card(key, pairing) {
    return <Card key={key} id={key} pairing={pairing} imageURL={this.props.imageURL(pairing)} checkMatch={this.checkMatch.bind(this)} matchedPairs={this.props.matchedPairs} activeCards={this.props.activeCards} />;
  }

  cards() {
    let cards = [];
    let key = 0;
    for(let i = 0; i < this.props.count; i++) {
      cards.push(this.card(key, i));
      key++;
      cards.push(this.card(key, i));
      key++;
    }

    cards.sort((a, b) => {
      const aPos = this.state.order.indexOf(a.props.id);
      const bPos = this.state.order.indexOf(b.props.id);
      if(aPos < bPos) {
        return -1;
      }
      if(aPos > bPos) {
        return 1;
      }
      return 0;
    });

    return cards;
  }

  shuffle(elements) {
    let temp, j, i = elements.length;
    while(--i) {
      j = Math.floor(Math.random() * (i + 1));
      temp = elements[i];
      elements[i] = elements[j];
      elements[j] = temp;
    }
    return elements;
  }

  finishedElement() {
    if(this.props.finished) {
      return <h1>You won!</h1>;
    } else {
      return null;
    }
  }

  classNames() {
    let classes  = ["cardList"];
    if(this.props.finished) {
      classes.push('-finished');
    }
    return classes.join(' ');
  }

  render() {
    return (
      <div className={this.classNames()}>
        {this.finishedElement()}
        {this.cards()}
      </div>
    );
  }
}

export default CardList;
