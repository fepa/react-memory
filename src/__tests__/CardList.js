import React from 'react';
import ReactDOM from 'react-dom';
import CardList from '../CardList';
import renderer from 'react-test-renderer';

const instance = <CardList
                   finished={false}
                   activeCards={[]}
                   count={1}
                   imageURL={() => { return 'foo'; }}
                   matchedPairs={{}}
                 />;

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(instance, div);
});

it('does not render the finished header by default', () => {
  expect(renderer.create(instance).toJSON()).toMatchSnapshot();
});

it('renders the finished header when all pairs have been found', () => {
  const component = renderer.create(
    <CardList
      finished={true}
      activeCards={[]}
      count={1}
      imageURL={() => { return 'foo'; }}
      matchedPairs={{}}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});
