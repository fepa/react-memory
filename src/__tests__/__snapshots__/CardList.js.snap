// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render the finished header by default 1`] = `
<div
  className="cardList"
>
  <div
    className="card -face-down"
    onClick={[Function]}
  >
    <img
      alt="Card"
      src="foo"
    />
  </div>
  <div
    className="card -face-down"
    onClick={[Function]}
  >
    <img
      alt="Card"
      src="foo"
    />
  </div>
</div>
`;

exports[`renders the finished header when all pairs have been found 1`] = `
<div
  className="cardList -finished"
>
  <h1>
    You won!
  </h1>
  <div
    className="card -face-down"
    onClick={[Function]}
  >
    <img
      alt="Card"
      src="foo"
    />
  </div>
  <div
    className="card -face-down"
    onClick={[Function]}
  >
    <img
      alt="Card"
      src="foo"
    />
  </div>
</div>
`;
